src/entities/join/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,187 @@
+import { encodeFunctionData } from 'viem';
+import { Token } from '../../../entities/token';
+import { TokenAmount } from '../../../entities/tokenAmount';
+import { WeightedEncoder } from '../../../entities/encoders/weighted';
+import { Address } from '../../../types';
+import { BALANCER_VAULT, MAX_UINT256, ZERO_ADDRESS } from '../../../utils';
+import { vaultAbi } from '../../../abi';
+import {
+    BaseJoin,
+    JoinBuildOutput,
+    JoinInput,
+    JoinKind,
+    WeightedJoinQueryResult,
+    WeightedJoinCall,
+} from '../types';
+import { AmountsJoin, PoolState } from '../../types';
+import { doQueryJoin, getAmounts, parseJoinArgs } from '../../utils';
+
+export class WeightedJoin implements BaseJoin {
+    public async query(
+        input: JoinInput,
+        poolState: PoolState,
+    ): Promise<WeightedJoinQueryResult> {
+        const amounts = this.getAmountsQuery(poolState.tokens, input);
+
+        const userData = this.encodeUserData(input.kind, amounts);
+
+        const { args, tokensIn } = parseJoinArgs({
+            useNativeAssetAsWrappedAmountIn:
+                !!input.useNativeAssetAsWrappedAmountIn,
```

Comment:
@johngrantuk JoinInput type doesn't have this property "useNativeAssetAsWrappedAmountIn", should add this as a to do?